Orange/widgets/visualize/utils/widget.py

```diff
@@ -0,0 +1,623 @@
+    """Base class for projection widgets."""
+    class Inputs:
+        data = Input("Data", Table, default=True)
+        data_subset = Input("Data Subset", Table)
+
+    class Outputs:
+        selected_data = Output("Selected Data", Table, default=True)
+        annotated_data = Output(ANNOTATED_DATA_SIGNAL_NAME, Table)
+
+    class Warning(OWProjectionWidgetBase.Warning):
+        missing_coords = Msg(
+            "Plot cannot be displayed because data"
+            "is missing for all points")
+
+    class Error(OWProjectionWidgetBase.Error):
+        sparse_data = Msg("Sparse data is not supported")
+
+    settingsHandler = DomainContextHandler()
+    selection = Setting(None, schema_only=True)
+    auto_commit = Setting(True)
+
+    GRAPH_CLASS = OWScatterPlotBase
+    graph = SettingProvider(OWScatterPlotBase)
+    graph_name = "graph.plot_widget.plotItem"
+    embedding_variables_names = ("proj-x", "proj-y")
+
+    def __init__(self):
+        super().__init__()
+        self.subset_data = None
+        self.subset_indices = None
+        self._embedding_coords = None  # embeddings for valid data
+        self.__pending_selection = self.selection
+
+        # GUI
+        self._effects_box = None
+        self._plot_box = None
+        self.setup_gui()
+
+    @property
+    def embedding_valid_data(self):  # embedding for valid data
+        return self._embedding_coords
+
+    @property
+    def embedding_all_data(self):  # embedding for valid and invalid data
+        if self.data is None:
+            return None
+        embedding = np.zeros((len(self.data), 2), dtype=np.float)
+        embedding[self.valid_data] = self.embedding_valid_data
+        return embedding
+
+    # GUI
+    def setup_gui(self):
+        self._add_graph()
+        self._add_top_controls()
+        self._add_middle_controls()
+        self._add_bottom_controls()
+
+    def _add_graph(self):
+        box = gui.vBox(self.mainArea, True, margin=0)
+        self.graph = self.GRAPH_CLASS(self, box)
+        box.layout().addWidget(self.graph.plot_widget)
+
+    def _add_top_controls(self):
+        self.graph.gui.point_properties_box(self.controlArea)
+
+    def _add_middle_controls(self):
+        self._effects_box = self.graph.gui.effects_box(self.controlArea)
+        self._plot_box = self.graph.gui.plot_properties_box(self.controlArea)
+
+    def _add_bottom_controls(self):
+        self.controlArea.layout().addStretch(100)
+        self.graph.box_zoom_select(self.controlArea)
+        gui.auto_commit(self.controlArea, self, "auto_commit",
+                        "Send Selection", "Send Automatically")
+
+    # Input
+    @Inputs.data
+    @check_sql_input
+    def set_data(self, data):
+        self.closeContext()
+        self.clear()
+        self.data = data
+        self.check_data()
+        self.init_attr_values()
+        self.openContext(self.data)
+        self.cb_class_density.setEnabled(self.can_draw_density())
+
+    def check_data(self):
```

Comment:
I'm not sure any of these checks will fit any widget (Freeviz, maybe). I think the method should just call `clear_messages`.

- The parent already handles sparse data, except for coordinates. Scatter plot can (or should) handle sparse data for coordinates, too. Other widgets maybe as well. So the first check is (generally) not OK.
- Data is invalid if coordinates can't be computed. This depends upon the selection of attribute subset (e.g. x and y in scatter plot, variables in radviz...). This test is thus generally wrong. More importantly, computing `self.valid_data` here is misleading.